### ISSUE TITLE
Installation of SCIPoptsuite linking CPLEX

### ISSUE THREAD

**Issue by purin0604** [scipopt/PySCIPOpt#377]
Hi,

I am trying to install SCIPoptsuite with Cplex lp solver linked.
Is there a way to use pre-build installers with Cplex lp solver? 
I successfully installed scip-7.0.0.tgz with lps=cpx but it seems incompatible with PySCIPopt.
So I tried to install scipoptsuite-7.0.0.tgz with source code, but I keep having errors building with cmake. 
Given my current circumstance, is there any clever way to work this out?  

In short: I want to use SCIP with Cplex and PySCIPopt. What installation option is the best?

**Comment by mattmilten** [comment]
Hi!
All you need for PySCIPOpt is a libscip. The easiest way to build this is using CMake. In your case, you would need to configure your build as `cmake -Bbuild -H. -DLPS=cpx`. Maybe you also need to specify `-DCPLEX_DIR=<cplex/install>` if the CPLEX lib cannot be found automatically.

You absolutely need to build SCIP (and the libscip) yourself, because there is no pre-built installer for SCIP/cpx.

Cheers, Matthias

**Comment by purin0604** [comment]
Thanks for the answer.
Finally, I completed installing scipoptsuit with Cplex using cmake and Pyscipopt as well.
( For the future reference, Pyscipopt setup.py file doesn't seems to have the right path for some linking files (ex: config.h and blockmemshell/memory.h) which are generated during the suit installation. Either the path needs to be modified or the files can be copied into the pyscipopt src folder) 

However, I found that presolving is done differently when the model is built and ran with pyscipopt compared to solving it directly with SCIP interactive shell (by reading a model file).

After asking about this issue to SCIP mailing list, I received an answer that it could be because PySCIPOpt lacks of the PaPILO library. 
(The output indeed misses "MILP presolver" line when running the model built with pyscipopt)

Does it mean that I missed linking the PaPILO library when  I install PySCIPOpt? How can I expand the package to include PaPILO library and enables presolving for MILP?

**Comment by fserra** [comment]
it would be weird that the PaPILO library is missing. This would mean that you have two different SCIPs installed, one with PaPILO and another without. But maybe it happened, as you built SCIP with CPLEX, can you check whether you are using different SCIPs?

**Comment by purin0604** [comment]
I have installed only one SCIP with the following  cmake configuration.
` cmake -Bbuild -H. -DZLIB=off -DREADLINE=off -DZIMPL=off -DGMP=off -DLPSCHECK=off  -DLPS=cpx `

Also, I can find the PaPILO library is built in the binary folder of the SCIP that I have linked to PySCIPOpt. 

In fact, I tested pyscipopt by reading a .cip model file within PySCIPOpt to see how presolving process is done.
```
from pyscipopt import Model
scip = Model()
scip.readProblem("SCIP_MIP.cip")
scip.optimize()
```
And its output shows that it's using MILP presolver
```
original problem has 30203 variables (14903 bin, 0 int, 0 impl, 15300 cont) and 53426 constraints
presolving:
(round 1, fast)       19697 del vars, 19697 del conss, 0 add conss, 19459 chg bounds, 0 chg sides, 0 chg coeffs, 0 upgd conss, 0 impls, 3903 clqs
(round 2, fast)       21683 del vars, 40301 del conss, 0 add conss, 21445 chg bounds, 0 chg sides, 0 chg coeffs, 0 upgd conss, 0 impls, 3903 clqs
   (0.3s) running MILP presolver
   (0.3s) MILP presolver found nothing
(round 3, exhaustive) 21683 del vars, 40301 del conss, 0 add conss, 21445 chg bounds, 0 chg sides, 0 chg coeffs, 8334 upgd conss, 0 impls, 3903 clqs
(round 4, medium)     25543 del vars, 40301 del conss, 0 add conss, 21445 chg bounds, 3860 chg sides, 3860 chg coeffs, 8334 upgd conss, 4430 impls, 430 clqs
(round 5, fast)       25543 del vars, 43774 del conss, 0 add conss, 21445 chg bounds, 3860 chg sides, 3860 chg coeffs, 8334 upgd conss, 4430 impls, 430 clqs
   (0.4s) running MILP presolver
   (0.4s) MILP presolver found nothing
(round 6, exhaustive) 25543 del vars, 44161 del conss, 0 add conss, 21445 chg bounds, 3860 chg sides, 3860 chg coeffs, 8334 upgd conss, 4430 impls, 430 clqs
(round 7, exhaustive) 25635 del vars, 44161 del conss, 0 add conss, 21445 chg bounds, 3860 chg sides, 3860 chg coeffs, 8334 upgd conss, 4430 impls, 377 clqs
   (0.7s) probing cycle finished: starting next cycle
(round 8, exhaustive) 25635 del vars, 44161 del conss, 0 add conss, 21489 chg bounds, 3860 chg sides, 3860 chg coeffs, 8335 upgd conss, 73843 impls, 428 clqs
(round 9, fast)       25635 del vars, 44161 del conss, 0 add conss, 21489 chg bounds, 3904 chg sides, 3948 chg coeffs, 8335 upgd conss, 73843 impls, 428 clqs
   (0.7s) probing: 47/386 (12.2%) - 0 fixings, 0 aggregations, 69464 implications, 44 bound changes
   (0.7s) probing aborted: 50/50 successive totally useless probings
   Deactivated symmetry handling methods, since SCIP was built without symmetry detector (SYM=none).
presolving (10 rounds: 10 fast, 6 medium, 5 exhaustive):
 25635 deleted vars, 44161 deleted constraints, 0 added constraints, 21489 tightened bounds, 0 added holes, 3904 changed sides, 3948 changed coefficients
 73843 implications, 428 cliques
presolved problem has 4568 variables (386 bin, 0 int, 0 impl, 4182 cont) and 9265 constraints
   4431 constraints of type <varbound>
      1 constraints of type <knapsack>
     43 constraints of type <setppc>
   4790 constraints of type <linear>
transformed objective value is always integral (scale: 1)
Presolving Time: 0.69

```

The following is the output when I built the model with pyscipopt and run it.

```
presolving:
(round 1, fast)       9157 del vars, 9157 del conss, 0 add conss, 9157 chg bounds, 0 chg sides, 0 chg coeffs, 0 upgd conss, 0 impls, 27 clqs
(round 2, fast)       9157 del vars, 17473 del conss, 0 add conss, 9157 chg bounds, 0 chg sides, 0 chg coeffs, 0 upgd conss, 0 impls, 27 clqs
(round 3, exhaustive) 9157 del vars, 17473 del conss, 0 add conss, 9157 chg bounds, 0 chg sides, 0 chg coeffs, 6112 upgd conss, 0 impls, 27 clqs
   (0.6s) probing: 69/14878 (0.5%) - 0 fixings, 0 aggregations, 943 implications, 0 bound changes
   (0.6s) probing aborted: 50/50 successive totally useless probings
   Deactivated symmetry handling methods, since SCIP was built without symmetry detector (SYM=none).
presolving (4 rounds: 4 fast, 2 medium, 2 exhaustive):
 19459 deleted vars, 27775 deleted constraints, 0 added constraints, 9157 tightened bounds, 0 added holes, 0 changed sides, 0 changed coefficients
 7009 implications, 45 cliques
presolved problem has 10744 variables (4576 bin, 0 int, 0 impl, 6168 cont) and 25651 constraints
   6084 constraints of type <varbound>
      1 constraints of type <knapsack>
     27 constraints of type <setppc>
  19539 constraints of type <linear>
transformed objective value is always integral (scale: 1)
Presolving Time: 1.60
```

The presolving process is quite different and it does better job (quicker to solve optimality with the better presolved model ) when I read .cip file.  
Since I would like implement own branching rules and pricing problems, I want to get the equivalent level of presolving performance when I write the model within python-scip interface.

Any idea what could cause this difference in presolving performance?
I wonder if this difference in presolved model ( read model files vs build model with pyscipopt) is special issue happening to me or it is somewhat expected by the presolving algorithm even for the exactly same mathematical model. 
 
Thank you

**Comment by fserra** [comment]
interesting. can you share your model to try to reproduce this?
It could be the fact that maybe the variables are stored in a different order. To test this we would need to change the random seed, but this functionality is not wrapped in pyscipopt yet, afaik.

**Comment by purin0604** [comment]
I can share the model with you for the purpose of reproducing and fixing the problem, but I am afraid to share them here.
Can I send you an email?

**Comment by mattmilten** [comment]
Please share the start of the logs showing the two SCIP versions and all their linked-in components first, so we can be sure that you are indeed running the same SCIP in both setups.

**Comment by purin0604** [comment]
I first ran the model written in python and wrote problem to file 'SCIP_MIP_BP.cip'. The following is the start of the logs. 
```
wrote problem to file b'SCIP_MIP_BP.cip'
----------------read solution
---------------start optimizing
SCIP version 7.0.0 [precision: 8 byte] [memory: block] [mode: optimized] [LP solver: SoPlex 5.0.0] [GitHash: 0bc4dc9c65]
Copyright (C) 2002-2020 Konrad-Zuse-Zentrum fuer Informationstechnik Berlin (ZIB)
1/1 feasible solution given by solution candidate storage, new primal bound 5.000000e+00

IN  pricerinit
OUT  pricerinit
presolving:
(round 1, fast)       9141 del vars, 9141 del conss, 0 add conss, 9141 chg bounds, 0 chg sides, 0 chg coeffs, 0 upgd conss, 0 impls, 43 clqs
(round 2, fast)       9141 del vars, 17457 del conss, 0 add conss, 9141 chg bounds, 0 chg sides, 0 chg coeffs, 0 upgd conss, 0 impls, 43 clqs
(round 3, exhaustive) 9141 del vars, 17457 del conss, 0 add conss, 9141 chg bounds, 0 chg sides, 0 chg coeffs, 6128 upgd conss, 0 impls, 43 clqs
   (2.0s) probing: 69/72603 (0.1%) - 0 fixings, 0 aggregations, 943 implications, 0 bound changes
   (2.0s) probing aborted: 50/50 successive totally useless probings
   Deactivated symmetry handling methods, since SCIP was built without symmetry detector (SYM=none).
presolving (4 rounds: 4 fast, 2 medium, 2 exhaustive):
 9141 deleted vars, 17457 deleted constraints, 0 added constraints, 9141 tightened bounds, 0 added holes, 0 changed sides, 0 changed coefficients
 7009 implications, 61 cliques
presolved problem has 78771 variables (72603 bin, 0 int, 0 impl, 6168 cont) and 108033 constraints
   6084 constraints of type <varbound>
      1 constraints of type <knapsack>
     43 constraints of type <setppc>
 101905 constraints of type <linear>
Presolving Time: 2.00
```

Then, I read the file 'SCIP_MIP_BP.cip', which has been just written by above then solve in SCIP console. 


```
SCIP version 7.0.0 [precision: 8 byte] [memory: block] [mode: optimized] [LP solver: SoPlex 5.0.0] [GitHash: 0bc4dc9c65]
Copyright (C) 2002-2020 Konrad-Zuse-Zentrum fuer Informationstechnik Berlin (ZIB)

External codes:
  SoPlex 5.0.0         Linear Programming Solver developed at Zuse Institute Berlin (soplex.zib.de) [GitHash: 6535a3c8]
  CppAD 20180000.0     Algorithmic Differentiation of C++ algorithms developed by B. Bell (www.coin-or.org/CppAD)
  MPIR 3.0.0           Multiple Precision Integers and Rationals Library developed by W. Hart (mpir.org)
  ZIMPL 3.3.9          Zuse Institute Mathematical Programming Language developed by T. Koch (zimpl.zib.de)
  PaPILO 1.0.0         parallel presolve for integer and linear optimization (https://github.com/lgottwald/PaPILO) [GitHash: 471d13d]
  Ipopt 3.12.9         Interior Point Optimizer developed by A. Waechter et.al. (www.coin-or.org/Ipopt)

user parameter file <scip.set> not found - using default parameters
SCIP> read "D:\Box Sync\Research_Jungeun\Drone_EMS\SCIP\SCIP_MIP_BP.cip"

read problem <D:\Box Sync\Research_Jungeun\Drone_EMS\SCIP\SCIP_MIP_BP.cip>
============

original problem has 87912 variables (72612 bin, 0 int, 0 impl, 15300 cont) and 125490 constraints

SCIP> optimize

presolving:
(round 1, fast)       76522 del vars, 86632 del conss, 0 add conss, 65606 chg bounds, 0 chg sides, 0 chg coeffs, 0 upgd conss, 0 impls, 4742 clqs
(round 2, fast)       76522 del vars, 103264 del conss, 0 add conss, 65606 chg bounds, 0 chg sides, 0 chg coeffs, 0 upgd conss, 0 impls, 4742 clqs
   (1.0s) running MILP presolver
   (1.0s) MILP presolver (6 rounds): 58 aggregations, 544 fixings, 0 bound changes
(round 3, medium)     77124 del vars, 103264 del conss, 0 add conss, 65606 chg bounds, 0 chg sides, 0 chg coeffs, 0 upgd conss, 0 impls, 4742 clqs
(round 4, fast)       77124 del vars, 104410 del conss, 0 add conss, 65606 chg bounds, 0 chg sides, 0 chg coeffs, 0 upgd conss, 0 impls, 4758 clqs
(round 5, exhaustive) 77124 del vars, 104410 del conss, 0 add conss, 65606 chg bounds, 0 chg sides, 0 chg coeffs, 11157 upgd conss, 0 impls, 4758 clqs
(round 6, medium)     81762 del vars, 104413 del conss, 0 add conss, 65606 chg bounds, 4635 chg sides, 4635 chg coeffs, 11157 upgd conss, 6472 impls, 931 clqs
(round 7, fast)       81762 del vars, 108240 del conss, 0 add conss, 65606 chg bounds, 4635 chg sides, 4635 chg coeffs, 11157 upgd conss, 6472 impls, 933 clqs
   (1.0s) running MILP presolver
   (1.0s) MILP presolver (2 rounds): 3 aggregations, 0 fixings, 0 bound changes
(round 8, exhaustive) 81765 del vars, 109051 del conss, 3 add conss, 65606 chg bounds, 4635 chg sides, 4635 chg coeffs, 11157 upgd conss, 6472 impls, 935 clqs
   (2.0s) probing cycle finished: starting next cycle
   Deactivated symmetry handling methods, since SCIP was built without symmetry detector (SYM=none).
presolving (9 rounds: 9 fast, 5 medium, 3 exhaustive):
 81765 deleted vars, 109054 deleted constraints, 3 added constraints, 65606 tightened bounds, 0 added holes, 4635 changed sides, 4635 changed coefficients
 148614 implications, 1177 cliques
presolved problem has 6147 variables (523 bin, 0 int, 0 impl, 5624 cont) and 16439 constraints
   6472 constraints of type <varbound>
      1 constraints of type <knapsack>
     43 constraints of type <setppc>
   9923 constraints of type <linear>
transformed objective value is always integral (scale: 1)
Presolving Time: 2.00
  
```

**Comment by purin0604** [comment]
Is this difference caused by including Pricer  ?

**Comment by fserra** [comment]
I think it could be due to the fact that in one you are providing a feasible solution and the other one not.
